File: bindings/ts/test/feed.test.ts
// Change feed conformance (contract §3.10 / §6). Remote requires the server
// started with [feed] enabled; embedded requires feed-enabled config (this
// port's embedded connect does not), so embedded feed_tail/read → Unsupported
// and feed_shards is 1. node --test + bun test.

import { test, before, after } from "node:test";
import assert from "node:assert/strict";
import { connect, InvalidInputError, ProtocolError, UnsupportedError } from "../src/index.ts";
import { spawnServerConfig, uniqueMemUrl, s, type Server } from "./harness.ts";

let server: Server;
before(async () => {
  server = await spawnServerConfig("[feed]\nenabled = true\n", ["--threads", "1"]);
});
after(() => server?.close());

test("feed_shards / feed_tail / feed_read with frames + resume", async () => {
  const c = await connect(server.url);
  assert.ok((await c.feedShards()) >= 1);
  const tail = await c.feedTail(0);

  await c.set("fk1", "v1");
  await c.set("fk2", "v2");

  const batch = await c.feedRead(0, tail.generation, tail.nextOffset, null);
  assert.ok(batch.frames.length >= 2, "expected >= 2 frames");
  assert.ok(
    batch.frames.some((f) => f.argv.length > 0 && s(f.argv[0]!) === "SET"),
    "a SET frame is present",
  );

  const next = await c.feedRead(0, batch.generation, batch.nextOffset, null);
  assert.equal(next.frames.length, 0, "resume from tail is caught up");
  c.close();
});

test("stale cursor → Protocol (FEEDRESYNC) error", async () => {
  const c = await connect(server.url);
  await assert.rejects(
    c.feedRead(0, 999999, 0, null),
    (e) => e instanceof ProtocolError,
  );
  c.close();
});

test("embedded feed: shards=1, non-zero shard InvalidInput, disabled Unsupported", async () => {
  const c = await connect(uniqueMemUrl());
  assert.equal(await c.feedShards(), 1);
  await assert.rejects(c.feedTail(1), (e) => e instanceof InvalidInputError);
  await assert.rejects(c.feedTail(0), (e) => e instanceof UnsupportedError);
  c.close();
});

File: bindings/go/feed_test.go
package kevy

import (
	"testing"
)

// Change feed FEED.* (contract §3.10, §6).

func TestFeedReplayRemote(t *testing.T) {
	s := spawnServerConfig(t, "[feed]\nenabled = true\n", "--threads", "1")
	c := mustConnect(t, s.url())

	if n, _ := c.FeedShards(bg); n < 1 {
		t.Fatalf("FeedShards=%d", n)
	}
	gen, off, err := c.FeedTail(bg, 0)
	if err != nil {
		t.Fatalf("FeedTail: %v", err)
	}

	// Produce some effects.
	_ = c.Set(bg, b("fk1"), b("v1"))
	_ = c.Set(bg, b("fk2"), b("v2"))

	batch, err := c.FeedRead(bg, 0, gen, off, 0)
	if err != nil {
		t.Fatalf("FeedRead: %v", err)
	}
	if len(batch.Frames) < 2 {
		t.Fatalf("expected >=2 frames, got %d", len(batch.Frames))
	}
	// Frames carry the applied argv.
	sawSet := false
	for _, f := range batch.Frames {
		if len(f.Argv) > 0 && string(f.Argv[0]) == "SET" {
			sawSet = true
		}
	}
	if !sawSet {
		t.Fatalf("no SET frame in feed: %+v", batch.Frames)
	}

	// Resume from the returned cursor: catching up yields an empty batch.
	next, err := c.FeedRead(bg, 0, batch.Generation, batch.NextOffset, 0)
	if err != nil {
		t.Fatalf("FeedRead resume: %v", err)
	}
	if len(next.Frames) != 0 {
		t.Fatalf("resume from tail should be caught up, got %d frames", len(next.Frames))
	}
}

func TestFeedStaleCursorResync(t *testing.T) {
	s := spawnServerConfig(t, "[feed]\nenabled = true\n", "--threads", "1")
	c := mustConnect(t, s.url())
	// A wildly-stale generation should surface a FEEDRESYNC protocol error.
	_, err := c.FeedRead(bg, 0, 999999, 0, 0)
	if err == nil {
		t.Skip("server did not treat a stale generation as unservable")
	}
	if !IsKind(err, KindProtocol) {
		t.Fatalf("stale cursor: want Protocol, got %v", err)
	}
}

func TestFeedEmbeddedUnsupported(t *testing.T) {
	c := mustConnect(t, "mem://feed-bus")
	if n, _ := c.FeedShards(bg); n != 1 {
		t.Fatalf("embedded FeedShards=%d want 1", n)
	}
	// Non-zero shard → InvalidInput.
	if _, _, err := c.FeedTail(bg, 1); !IsKind(err, KindInvalidInput) {
		t.Fatalf("embedded non-zero shard: want InvalidInput, got %v", err)
	}
	// Feed not enabled on this port's embedded Connect → Unsupported.
	if _, _, err := c.FeedTail(bg, 0); !IsKind(err, KindUnsupported) {
		t.Fatalf("embedded feed disabled: want Unsupported, got %v", err)
	}
}
